src/persist-client/examples/open_loop.rs

```diff
@@ -173,50 +173,50 @@ where
                 // as a relative duration from start.
                 let mut prev_log = Duration::from_millis(0);
                 loop {
+                    // Data generation can be CPU expensive, so generate it
+                    // in a spawn_blocking to play nicely with the rest of
+                    // the async code.
+                    let mut data_generator = data_generator.clone();
+                    let batch = mz_ore::task::spawn_blocking(
+                        || "data_generator-batch",
+                        move || data_generator.gen_batch(usize::cast_from(batch_idx)),
+                    )
+                    .await
+                    .expect("task failed");
+                    trace!("data generator {} made a batch", idx);
+
+                    // Intentionally return before the sleep if we got a None
+                    // batch (indicating that we're done).
+                    let batch = match batch {
+                        Some(x) => x,
+                        None => {
+                            let records_sent = usize::cast_from(batch_idx) * args.batch_size;
+                            let finished = format!(
+                                "Data generator {} finished after {} ms and sent {} records",
+                                idx,
+                                start.elapsed().as_millis(),
+                                records_sent
+                            );
+                            return Ok(finished);
+                        }
+                    };
+                    batch_idx += 1;
+
                     // Sleep so this doesn't busy wait if it's ahead of
                     // schedule.
                     let elapsed = start.elapsed();
-                    let next_batch_time = time_per_batch * (batch_idx + 1);
+                    let next_batch_time = time_per_batch * (batch_idx);
                     let sleep = next_batch_time.saturating_sub(start.elapsed());
```

Comment:
really minor nit and I know this isn't part of your diff but you can reuse `elapsed` here instead of calling `start.elapsed()`

Reply:
done!